ideascube/serveradmin/wifi.py

```diff
@@ -0,0 +1,208 @@
+from collections import OrderedDict
+from itertools import groupby
+from operator import attrgetter
+from uuid import uuid4
+
+from django.utils.translation import ugettext as _
+
+try:
+    # The NetworkManager module tries to connect to the NetworkManager daemon
+    # right when we import it.
+    #
+    # Therefore, if there is no NetworkManager daemon running, or if it isn't
+    # reachable for some reason (like in our CI servers), the imports will
+    # fail.
+    from NetworkManager import (
+        NetworkManager,
+        Settings as NMSettings,
+        NM_DEVICE_TYPE_WIFI,
+        )
+
+except Exception as e:
+    # Set these to None so they exist, the code will handle exceptions.
+    #
+    # It also helps with unit tests, which wouldn't be able to mock them as
+    # easily if they didn't exist here.
+    NetworkManager = None
+    NMSettings = None
+
+    # Hardcode this one, it is a constant anyway in the NetworkManager module,
+    # and hardcoding it allows the tests to run no matter what.
+    NM_DEVICE_TYPE_WIFI = 2
+
+
+class AvailableWifiNetwork(object):
+    @classmethod
+    def all(cls):
+        result = []
+
+        known_connections = KnownWifiConnection.all()
+
+        device = get_wifi_device()
+        aps = device.SpecificDevice().GetAllAccessPoints()
+        aps.sort(key=attrgetter('Frequency'), reverse=True)
+        aps.sort(key=attrgetter('Ssid'))
+
+        for ssid, grouper in groupby(aps, attrgetter('Ssid')):
+            if ssid:
+                # There often are more than one network for a given SSID (for
+                # example at 2.4 and 5GHz), so we keep only the one with the
+                # highest frequency.
+                ap = next(grouper)
+                result.append(cls(ap, device, known_connections))
+
+        return OrderedDict(
+            [(n.ssid, n) for n in sorted(
+                result, key=attrgetter('strength'), reverse=True)])
+
+    def __init__(self, access_point, device, known_connections=None):
+        self._access_point = access_point
+        self._device = device
+
+        known_connections = known_connections if known_connections else {}
+        self._connection = known_connections.get(self.ssid, None)
```

Comment:
I know, I add it out of habit, to be entirely explicit on the expected default value.